core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -598,20 +649,39 @@ public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec update
       builder.add(freshSpec);
     }
 
+    // determine the next order id
+    int maxOrderId = sortOrders.stream().mapToInt(SortOrder::orderId).max().orElse(INITIAL_SORT_ORDER_ID);
+    int nextOrderId = maxOrderId + 1;
+
+    // rebuild the sort order using new column ids
+    SortOrder freshSortOrder = freshSortOrder(nextOrderId, freshSchema, updatedSortOrder);
+
+    // if the order already exists, use the same ID. otherwise, use 1 more than the highest ID.
+    Optional<SortOrder> sameSortOrder = sortOrders.stream()
+        .filter(sortOrder -> sortOrder.sameOrder(freshSortOrder))
+        .findAny();
+    int orderId = sameSortOrder.map(SortOrder::orderId).orElse(nextOrderId);
+
+    ImmutableList.Builder<SortOrder> sortOrdersBuilder = ImmutableList.<SortOrder>builder().addAll(sortOrders);
```

Comment:
This place, probably, requires additional attention. I've added a test to `TestReplaceTransaction`.